### PR TITLE
CXX-518 don't allow sending _id with nested $ key

### PR DESCRIPTION
**Pull request by amidvidy**

### CODE REVIEW

File: src/mongo/SConscript [hunk 1]
```diff
@@ -53,14 +53,15 @@ unittests = [
     'base/parse_number_test',
     'bson/bson_field_test',
     'bson/bson_obj_test',
```
Comment (amidvidy) [inline src/mongo/SConscript]: extra changes are from sorting these lines

File: src/mongo/client/insert_write_operation.cpp [hunk 2]
```diff
@@ -66,8 +66,10 @@ namespace mongo {
     }
 
     BSONObj InsertWriteOperation::_ensureId(const BSONObj& doc) {
-        if (doc.hasField("_id"))
+        if (doc.hasField("_id")) {
+            uassert(0, "value of _id element cannot contain any fields starting with $", doc.okForStorage());
```
Comment (acmorrow) [inline src/mongo/client/insert_write_operation.cpp]: Validating the whole object seems excessive. The point of this is to protect old servers from a certain unchecked failure mode, but only w.r.t. to the _id element.

I think you should re-write this, more like:

```
BSONElement id = doc.getField("_id");
if (!id.eoo()) {
    if (id has substructure) {
        BSONObj idObj = // get element as object
        uassert(0, "...", idObj.okForStorage();
    }
}
```

Comment (amidvidy) [inline src/mongo/client/insert_write_operation.cpp]: done.

File: src/mongo/client/insert_write_operation_test.cpp [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        bob.append("foo", "bar");
+        const BSONObj doc = bob.done();
+        ASSERT_THROWS(InsertWriteOperation w(doc), UserException);
+    }
+
+    TEST(InsertWriteOperation, IdFieldCannotContainDollarSignNested) {
+        BSONObjBuilder bob;
+        BSONObjBuilder subbob;
+        BSONObjBuilder subsubbob;
+        BSONObjBuilder subsubsubbob;
```
Comment (acmorrow) [inline src/mongo/client/insert_write_operation_test.cpp]: This is a somewhat odd way of using BSONObjBuilder. Using 'subobjStart' might work better.

File: src/mongo/client/insert_write_operation.cpp [hunk 4]
```diff
@@ -66,8 +67,11 @@ namespace mongo {
     }
 
     BSONObj InsertWriteOperation::_ensureId(const BSONObj& doc) {
-        if (doc.hasField("_id"))
+        BSONElement id = doc.getField("_id");
+        if (!id.eoo()) {
+            uassert(0, "value of _id element cannot contain any fields starting with $", !id.isABSONObj() || id.Obj().okForStorage());
```
Comment (acmorrow) [inline src/mongo/client/insert_write_operation.cpp]: Does isABSONObj also handle the case where _id is an array (I don't know that is possible, btw).

Comment (amidvidy) [inline src/mongo/client/insert_write_operation.cpp]: yes, it does.

File: src/mongo/client/insert_write_operation_test.cpp (hunk 3 above)
Comment (amidvidy) [inline src/mongo/client/insert_write_operation_test.cpp]: my bad, I've been spoiled by `bsoncxx`. fixed

**Comment by acmorrow** [comment]
LGTM

**Pull request merged**